diff --git a/include/boost/asio/signal_set_base.hpp b/include/boost/asio/signal_set_base.hpp
index 29161ade..1019e6c4 100644
--- a/include/boost/asio/signal_set_base.hpp
+++ b/include/boost/asio/signal_set_base.hpp
@@ -61,7 +61,12 @@ public:
   enum class flags : int
   {
     none = 0,
+    // SA_RESTART is not supported in QNX
+#ifdef __QNX__
+    restart = -1,
+#else
     restart = BOOST_ASIO_OS_DEF(SA_RESTART),
+#endif
     no_child_stop = BOOST_ASIO_OS_DEF(SA_NOCLDSTOP),
     no_child_wait = BOOST_ASIO_OS_DEF(SA_NOCLDWAIT),
     dont_care = -1
