From 7ba2420162b69a801f175a9981c019253c68c9a3 Mon Sep 17 00:00:00 2001
From: Yongxin Dai <yodai@blackberry.com>
Date: Fri, 17 Sep 2021 13:01:10 -0400
Subject: [PATCH] Support QNX platform

---
 BUILD.gn | 4 ++++
 1 file changed, 4 insertions(+)

diff --git a/BUILD.gn b/BUILD.gn
index bceab17c..89af31e7 100644
--- a/BUILD.gn
+++ b/BUILD.gn
@@ -242,6 +242,10 @@ if (libyuv_use_neon) {
       configs -= [ "//build/config/compiler:compiler_arm_fpu" ]
       cflags = [ "-mfpu=neon" ]
     }
+
+    if ( current_cpu == "arm64" && is_qnx ) {
+      cflags_cc = [ "-flax-vector-conversions" ]
+    }
   }
 }
 
-- 
2.34.1

