diff --git a/src/find.c b/src/find.c
index d5dbf7a..cd2728e 100644
--- a/src/find.c
+++ b/src/find.c
@@ -60,7 +60,7 @@ rcutils_find_lastn(const char * str, char delimiter, size_t string_length)
     return SIZE_MAX;
   }
 
-#if defined(_GNU_SOURCE)
+#if defined(_GNU_SOURCE) && !defined(__QNX__)
   const char * ptr = memrchr(str, delimiter, string_length);
   if (ptr == NULL) {
     return SIZE_MAX;
