diff --git a/rosidl_generator_cpp/cmake/rosidl_generator_cpp_generate_interfaces.cmake b/rosidl_generator_cpp/cmake/rosidl_generator_cpp_generate_interfaces.cmake
index 864ed78..99a5339 100644
--- a/rosidl_generator_cpp/cmake/rosidl_generator_cpp_generate_interfaces.cmake
+++ b/rosidl_generator_cpp/cmake/rosidl_generator_cpp_generate_interfaces.cmake
@@ -138,7 +138,9 @@ add_dependencies(
 
 set(_target_suffix "__rosidl_generator_cpp")
 add_library(${rosidl_generate_interfaces_TARGET}${_target_suffix} INTERFACE)
-target_compile_features(${rosidl_generate_interfaces_TARGET}${_target_suffix} INTERFACE cxx_std_17)
+if(NOT QNX)
+  target_compile_features(${rosidl_generate_interfaces_TARGET}${_target_suffix} INTERFACE cxx_std_17)
+endif()
 add_library(${PROJECT_NAME}::${rosidl_generate_interfaces_TARGET}${_target_suffix} ALIAS
   ${rosidl_generate_interfaces_TARGET}${_target_suffix})
 target_include_directories(${rosidl_generate_interfaces_TARGET}${_target_suffix}
diff --git a/rosidl_runtime_c/CMakeLists.txt b/rosidl_runtime_c/CMakeLists.txt
index 825a026..c56bfe4 100644
--- a/rosidl_runtime_c/CMakeLists.txt
+++ b/rosidl_runtime_c/CMakeLists.txt
@@ -65,6 +65,10 @@ ament_export_targets(${PROJECT_NAME})
 
 ament_index_register_resource("rosidl_runtime_packages")
 
+if(QNX AND QNX_TESTING)
+  set(BUILD_TESTING ON)
+endif()
+
 if(BUILD_TESTING)
   find_package(ament_lint_auto REQUIRED)
   find_package(performance_test_fixture REQUIRED)
@@ -151,6 +155,10 @@ if(BUILD_TESTING)
   if(TARGET benchmark_string_conversion)
     target_link_libraries(benchmark_string_conversion ${PROJECT_NAME})
   endif()
+
+  if(QNX)
+    install(DIRECTORY ${CMAKE_BINARY_DIR}/ DESTINATION test/rosidl_runtime_c)
+  endif()
 endif()
 
 install(
