From 478f53587ead63220e31a674256bc1fe39432d77 Mon Sep 17 00:00:00 2001
From: Yongxin Dai <yodai@qnx.com>
Date: Mon, 26 May 2025 08:31:46 -0400
Subject: [QNX LOCAL PATCH] include tests for QNX

---
 webrtc.gni | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/webrtc.gni b/webrtc.gni
index 709d8416c8..6a09451900 100644
--- a/webrtc.gni
+++ b/webrtc.gni
@@ -310,7 +310,7 @@ declare_args() {
 
   # Set this to true to build the unit tests.
   # Disabled when building with Chromium or Mozilla.
-  rtc_include_tests = !build_with_chromium && !build_with_mozilla && !is_qnx
+  rtc_include_tests = !build_with_chromium && !build_with_mozilla
 
   # Set this to false to skip building code that also requires X11 extensions
   # such as Xdamage, Xfixes.
-- 
2.34.1

