From 04b71807f211c3e367dc8e4bcb7219b684783333 Mon Sep 17 00:00:00 2001
From: Yongxin Dai <yodai@blackberry.com>
Date: Sun, 9 Feb 2025 16:46:00 -0500
Subject: [QNX LOCAL PATCH] include ifaddrs.h for QNX

---
 rtc_base/net_test_helpers.cc | 3 +++
 1 file changed, 3 insertions(+)

diff --git a/rtc_base/net_test_helpers.cc b/rtc_base/net_test_helpers.cc
index 806d7dee60..d174a43b44 100644
--- a/rtc_base/net_test_helpers.cc
+++ b/rtc_base/net_test_helpers.cc
@@ -23,6 +23,9 @@
 #include <arpa/inet.h>
 #if defined(WEBRTC_ANDROID)
 #include "rtc_base/ifaddrs_android.h"
+#elif defined(WEBRTC_QNX)
+#include <ifaddrs.h>
+#include <sys/socket.h>
 #else
 #include <ifaddrs.h>
 #endif
-- 
2.34.1

