From df50995de06ec6200a6088dc2e0a77d6a2e3ed17 Mon Sep 17 00:00:00 2001
From: Yongxin Dai <yodai@qnx.com>
Date: Mon, 17 Mar 2025 09:25:43 -0400
Subject: [QNX LOCAL PATCH] extend the default timeout period for QNX.

---
 pc/peer_connection_encodings_integrationtest.cc | 4 ++++
 1 file changed, 4 insertions(+)

diff --git a/pc/peer_connection_encodings_integrationtest.cc b/pc/peer_connection_encodings_integrationtest.cc
index 6ebc247a46..dfd105816d 100644
--- a/pc/peer_connection_encodings_integrationtest.cc
+++ b/pc/peer_connection_encodings_integrationtest.cc
@@ -66,7 +66,11 @@ namespace webrtc {
 
 namespace {
 
+#if defined(WEBRTC_QNX)
+constexpr TimeDelta kDefaultTimeout = TimeDelta::Seconds(12);
+#else
 constexpr TimeDelta kDefaultTimeout = TimeDelta::Seconds(5);
+#endif
 // Most tests pass in 20-30 seconds, but some tests take longer such as AV1
 // requiring additional ramp-up time (https://crbug.com/webrtc/15006) or SVC
 // (LxTx_KEY) being slower than simulcast to send top spatial layer.
-- 
2.34.1

