diff --git a/SAX2.c b/SAX2.c
index b5c2e476..21388870 100644
--- a/SAX2.c
+++ b/SAX2.c
@@ -427,8 +427,13 @@ xmlSAX2ResolveEntity(void *ctx, const xmlChar *publicId,
             if (res < 0)
                 xmlSAX2ErrMemory(ctxt);
             else
+#ifdef __QNX__
+            xmlWarnMsg(ctxt, XML_ERR_INVALID_URI,
+                           "Can't resolve URI: %s\n", systemId==NULL?" ":systemId);
+#else
                 xmlWarnMsg(ctxt, XML_ERR_INVALID_URI,
                            "Can't resolve URI: %s\n", systemId);
+#endif
             return(NULL);
         }
         if (xmlStrlen(URI) > XML_MAX_URI_LENGTH) {
diff --git a/parser.c b/parser.c
index bdc0aae1..f0123bb4 100644
--- a/parser.c
+++ b/parser.c
@@ -250,6 +250,7 @@ static void LIBXML_ATTR_FORMAT(3,0)
 xmlFatalErrMsg(xmlParserCtxtPtr ctxt, xmlParserErrors error,
                const char *msg)
 {
+    if(msg == NULL) msg = "";
     xmlCtxtErr(ctxt, NULL, XML_FROM_PARSER, error, XML_ERR_FATAL,
                NULL, NULL, NULL, 0, "%s", msg);
 }
