diff --git a/rosidl_generator_py/cmake/rosidl_generator_py_generate_interfaces.cmake b/rosidl_generator_py/cmake/rosidl_generator_py_generate_interfaces.cmake
index a6d8263..12550ba 100644
--- a/rosidl_generator_py/cmake/rosidl_generator_py_generate_interfaces.cmake
+++ b/rosidl_generator_py/cmake/rosidl_generator_py_generate_interfaces.cmake
@@ -176,6 +176,11 @@ target_link_libraries(
   ${_target_name_lib}
   ${PythonExtra_LIBRARIES}
 )
+
+if(QNX)
+  list(APPEND PythonExtra_INCLUDE_DIRS ${CMAKE_INSTALL_PREFIX}/usr/lib/python${PYTHON_VERSION_QNX}/site-packages/numpy/core/include)
+endif()
+
 target_include_directories(${_target_name_lib}
   PRIVATE
   ${CMAKE_CURRENT_BINARY_DIR}/rosidl_generator_c
@@ -188,7 +193,7 @@ find_file(_numpy_h numpy/numpyconfig.h
   PATHS ${PythonExtra_INCLUDE_DIRS}
 )
 
-if(APPLE OR WIN32 OR NOT _numpy_h)
+if(NOT QNX AND (APPLE OR WIN32 OR NOT _numpy_h))
   # add include directory for numpy headers
   set(_python_code
     "import numpy"
@@ -209,6 +214,10 @@ if(APPLE OR WIN32 OR NOT _numpy_h)
   target_include_directories(${_target_name_lib} PUBLIC "${_output}")
 endif()
 
+if(QNX)
+  target_include_directories(${_target_name_lib} PUBLIC ${QNX_TARGET}/${CPUVARDIR}/opt/ros/humble/usr/lib/python3.11/site-packages/numpy/core/include)
+endif()
+
 rosidl_get_typesupport_target(c_typesupport_target "${rosidl_generate_interfaces_TARGET}" "rosidl_typesupport_c")
 target_link_libraries(${_target_name_lib} ${c_typesupport_target})
 
