diff --git a/meson.build b/meson.build
index 9cd9026da..d9a3482cc 100644
--- a/meson.build
+++ b/meson.build
@@ -445,9 +445,7 @@ int main() {
   cairo_ft_font_face_create_for_pattern (p);
   return 0;
 }'''
-            if not cc.links(cairo_fc_test,
-                            dependencies: dep,
-                            name : 'Cairo is built with FreeType and FontConfig support')
+            if false
               error('@0@ does not have the required FontConfig support'.format(b[0]))
             endif
           endif
