From cc9ef9459d0439b778863953b2258c1b584d5f65 Mon Sep 17 00:00:00 2001
From: Yongxin Dai <yodai@blackberry.com>
Date: Sat, 15 Apr 2023 12:57:30 -0400
Subject: [PATCH] Support QNX platform

---
 aom_dsp/arm/mem_neon.h | 8 ++++++++
 1 file changed, 8 insertions(+)

diff --git a/aom_dsp/arm/mem_neon.h b/aom_dsp/arm/mem_neon.h
index 92ad76960d..a223c21ddc 100644
--- a/aom_dsp/arm/mem_neon.h
+++ b/aom_dsp/arm/mem_neon.h
@@ -65,11 +65,19 @@ static INLINE uint16x8x4_t vld1q_u16_x4(const uint16_t *ptr) {
 #endif  // __GNUC__ < 8
 
 #if __GNUC__ < 9
+#if defined(__QNXNTO__)
+static INLINE uint16x8x4_t vld1q_u16_x4(const uint16_t *ptr) {
+  uint16x8x4_t res = { { vld1q_u16(ptr + 0 * 8), vld1q_u16(ptr + 1 * 8),
+                         vld1q_u16(ptr + 2 * 8), vld1q_u16(ptr + 3 * 8) } };
+  return res;
+}
+#else
 static INLINE uint8x16x3_t vld1q_u8_x3(const uint8_t *ptr) {
   uint8x16x3_t res = { { vld1q_u8(ptr + 0 * 16), vld1q_u8(ptr + 1 * 16),
                          vld1q_u8(ptr + 2 * 16) } };
   return res;
 }
+#endif
 #endif  // __GNUC__ < 9
 #endif  // defined(__GNUC__) && !defined(__clang__)
 
-- 
2.34.1

