From 546f0783eb4d64964a3055b98cb4fe795e7bdb27 Mon Sep 17 00:00:00 2001
From: Yongxin Dai <yodai@qnx.com>
Date: Wed, 2 Jul 2025 15:45:18 -0400
Subject: [PATCH] Fix 'using iterator_concept' for SDP7

---
 include/v8-internal.h | 4 ++++
 1 file changed, 4 insertions(+)

diff --git a/include/v8-internal.h b/include/v8-internal.h
index f81a2108f34..cd3c8cf7bed 100644
--- a/include/v8-internal.h
+++ b/include/v8-internal.h
@@ -1432,7 +1432,11 @@ struct MaybeDefineIteratorConcept {};
 template <typename Iterator>
 struct MaybeDefineIteratorConcept<
     Iterator, std::enable_if_t<kHaveIteratorConcept<Iterator>>> {
+#if !defined(__QNX__) || __QNX__ >= 800
   using iterator_concept = Iterator::iterator_concept;
+#else
+  typename Iterator::iterator_concept iterator_concept;
+#endif
 };
 // Otherwise fall back to `std::iterator_traits<Iterator>` if possible.
 template <typename Iterator>
-- 
2.34.1

