From 4fabdc104fb0844aa0a01d86e39be90e0b9b14a1 Mon Sep 17 00:00:00 2001
From: Yongxin Dai <yodai@qnx.com>
Date: Sun, 6 Jul 2025 09:31:19 -0400
Subject: [PATCH] Add 'typename' to satisfy SDP7

---
 src/sandbox/external-entity-table.h | 13 +++++++++++++
 1 file changed, 13 insertions(+)

diff --git a/src/sandbox/external-entity-table.h b/src/sandbox/external-entity-table.h
index 5928cf61a61..a0fa6f5b91d 100644
--- a/src/sandbox/external-entity-table.h
+++ b/src/sandbox/external-entity-table.h
@@ -50,10 +50,23 @@ class V8_EXPORT_PRIVATE ExternalEntityTable
     : public SegmentedTable<Entry, size> {
  protected:
   using Base = SegmentedTable<Entry, size>;
+#if !defined(__QNX__) || __QNX__ >= 800
   using FreelistHead = Base::FreelistHead;
+#else
+  using FreelistHead =  typename SegmentedTable<Entry, size>::FreelistHead;
+#endif
+#if !defined(__QNX__) || __QNX__ >= 800
   using Segment = Base::Segment;
+#else
+  using Segment = typename SegmentedTable<Entry, size>::Segment;
+#endif
+#if !defined(__QNX__) || __QNX__ >= 800
   using WriteIterator = Base::WriteIterator;
+#else
+  using WriteIterator = typename SegmentedTable<Entry, size>::WriteIterator;
+#endif
   static constexpr size_t kSegmentSize = Base::kSegmentSize;
+
   static constexpr size_t kEntriesPerSegment = Base::kEntriesPerSegment;
   static constexpr size_t kEntrySize = Base::kEntrySize;
 
-- 
2.34.1

