From 028df704e28d6a0548660f4e4be52c0f0f94fae1 Mon Sep 17 00:00:00 2001
From: Yongxin Dai <yodai@qnx.com>
Date: Tue, 8 Jul 2025 08:02:10 -0400
Subject: [PATCH] constxpr operator T() for SDP7

---
 src/flags/flags.h | 5 ++++-
 1 file changed, 4 insertions(+), 1 deletion(-)

diff --git a/src/flags/flags.h b/src/flags/flags.h
index 70f84bf84da..400e8c136d6 100644
--- a/src/flags/flags.h
+++ b/src/flags/flags.h
@@ -44,8 +44,11 @@ class FlagValue {
 
   // Implicitly convert to a {T}. Not marked {constexpr} so we do not get
   // compiler warnings about dead code (when checking readonly flags).
+ #if !defined(__QNX__) || __QNX__ >= 800
   operator T() const { return value_; }
-
+#else
+  constexpr operator T() const { return value_; }
+#endif
   // Explicitly convert to a {T} via {value()}. This is {constexpr} so we can
   // use it for computing other constants.
   constexpr T value() const { return value_; }
-- 
2.34.1

