From f247edcbc0706902f2cc92e1bc1bed8efc01688e Mon Sep 17 00:00:00 2001
From: Yongxin Dai <yodai@qnx.com>
Date: Tue, 8 Jul 2025 15:46:50 -0400
Subject: [PATCH] Disable one static_assert for SDP7

---
 src/profiler/heap-snapshot-generator.cc | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/src/profiler/heap-snapshot-generator.cc b/src/profiler/heap-snapshot-generator.cc
index f39f1cb1e09..48a19cb5378 100644
--- a/src/profiler/heap-snapshot-generator.cc
+++ b/src/profiler/heap-snapshot-generator.cc
@@ -879,10 +879,12 @@ struct ManagedName {
 };
 
 // A little inline test:
+#if !defined(__QNX__) || __QNX__ >= 800
 constexpr const char kTagNameForTesting[] = "kFooTag";
 static_assert(std::string_view{
                   ManagedName<kTagNameForTesting>::str_arr.data()} ==
               std::string_view{"system / Managed<Foo>"});
+#endif
 }  // namespace
 
 HeapEntry* V8HeapExplorer::AddEntry(Tagged<HeapObject> object) {
-- 
2.34.1

