diff --git a/src/CMakeLists.txt b/src/CMakeLists.txt
index f59467c0..547bfe28 100644
--- a/src/CMakeLists.txt
+++ b/src/CMakeLists.txt
@@ -538,13 +538,19 @@ if(WITH_GNUTLS)
 endif()
 
 if(WITH_PCRE2)
-	pkg_check_modules(PCRE2 libpcre2-8)
+	find_package(PCRE2 REQUIRED 8BIT CONFIG)
 	if(PCRE2_FOUND)
-		set(PCRE_LDFLAGS "${PCRE2_LDFLAGS}")
+		set(PCRE_LDFLAGS PCRE2::8BIT)
 		set(PCRE_CFLAGS "${PCRE2_CFLAGS}")
 	else()
-		## if we have pcre2-config, use it
-		xconfig(pcre2-config PCRE_LDFLAGS PCRE_CFLAGS)
+		pkg_check_modules(PCRE2 libpcre2-8)
+		if(PCRE2_FOUND)
+			set(PCRE_LDFLAGS "${PCRE2_LDFLAGS}")
+			set(PCRE_CFLAGS "${PCRE2_CFLAGS}")
+		else()
+			## if we have pcre2-config, use it
+			xconfig(pcre2-config PCRE_LDFLAGS PCRE_CFLAGS)
+		endif()
 	endif()
 	if(PCRE_LDFLAGS OR PCRE_CFLAGS)
 		message(STATUS "found pcre2 at: LDFLAGS: ${PCRE_LDFLAGS} CFLAGS: ${PCRE_CFLAGS}")
