diff --git a/test/CMakeLists.txt b/test/CMakeLists.txt
index edd90a1..04a96ac 100644
--- a/test/CMakeLists.txt
+++ b/test/CMakeLists.txt
@@ -34,10 +34,19 @@ function(add_coverage_test Name)
         PRIVATE
         GTest::gtest_main
         GTest::gmock
-        pthread
     )
     target_include_directories(${Name} PRIVATE ${CMAKE_CURRENT_SOURCE_DIR}/include)
-    gtest_discover_tests(${Name} XML_OUTPUT_DIR results)
+    if(CMAKE_SYSTEM_NAME MATCHES "QNX")
+        # QNX provides pthread inside libc and does not need to link pthread lib.
+        #
+        # For QNX we are using cross compilation.
+        # Thus, we can't call gtest_discover_tests()
+        # Instead, we call old gtest_add_tests()
+        gtest_add_tests(TARGET ${Name} SOURCES ${ARGN})
+    else()
+        target_link_libraries(${Name} PRIVATE pthread)
+        gtest_discover_tests(${Name} XML_OUTPUT_DIR results)
+    endif()
 endfunction()
 
 # NOTE: This is temporarily just a call to add_coverage_test. When coverage
