diff --git a/test/CMakeLists.txt b/test/CMakeLists.txt
index 3c9a6f5..ea202a0 100644
--- a/test/CMakeLists.txt
+++ b/test/CMakeLists.txt
@@ -35,10 +35,19 @@ function(add_coverage_test Name)
         PRIVATE
         GTest::gtest_main
         GTest::gmock
-        pthread
     )
     target_include_directories(${Name} PRIVATE ${CMAKE_CURRENT_SOURCE_DIR}/include)
-    gtest_discover_tests(${Name} XML_OUTPUT_DIR results)
+    if(CMAKE_SYSTEM_NAME MATCHES "QNX")
+        # QNX provides pthread inside libc and does not need to link pthread lib.
+        #
+        # For QNX we are using cross compilation.
+        # Thus, we can't call gtest_discover_tests()
+        # Instead, we call old gtest_add_tests()
+        gtest_add_tests(TARGET ${Name} SOURCES ${ARGN})
+    else()
+        target_link_libraries(${Name} PRIVATE pthread)
+        gtest_discover_tests(${Name} XML_OUTPUT_DIR results)
+    endif()
 endfunction()
 
 # NOTE: This is temporarily just a call to add_coverage_test. When coverage
diff --git a/test/coverage/communication/RpcClientTest.cpp b/test/coverage/communication/RpcClientTest.cpp
index 94d7e79..c30d8c7 100644
--- a/test/coverage/communication/RpcClientTest.cpp
+++ b/test/coverage/communication/RpcClientTest.cpp
@@ -35,6 +35,6 @@ protected:
 };
 
 // TODO replace
-TEST_F(TestFixture, SomeTestName) {}
+TEST_F(TestFixture, SomeTestName3) {}
 
 }  // namespace
diff --git a/test/coverage/utils/CyclicQueueTest.cpp b/test/coverage/utils/CyclicQueueTest.cpp
index 8160e8b..5152dfc 100644
--- a/test/coverage/utils/CyclicQueueTest.cpp
+++ b/test/coverage/utils/CyclicQueueTest.cpp
@@ -33,6 +33,6 @@ protected:
 };
 
 // TODO replace
-TEST_F(TestFixture, SomeTestName) {}
+TEST_F(TestFixture, SomeTestName5) {}
 
 }  // namespace
diff --git a/test/coverage/utils/IpAddressTest.cpp b/test/coverage/utils/IpAddressTest.cpp
index b243bd0..0dfed14 100644
--- a/test/coverage/utils/IpAddressTest.cpp
+++ b/test/coverage/utils/IpAddressTest.cpp
@@ -33,6 +33,6 @@ protected:
 };
 
 // TODO replace
-TEST_F(TestFixture, SomeTestName) {}
+TEST_F(TestFixture, SomeTestName6) {}
 
 }  // namespace
diff --git a/test/coverage/utils/ThreadPoolTest.cpp b/test/coverage/utils/ThreadPoolTest.cpp
index 22b1dfd..d29608a 100644
--- a/test/coverage/utils/ThreadPoolTest.cpp
+++ b/test/coverage/utils/ThreadPoolTest.cpp
@@ -33,6 +33,6 @@ protected:
 };
 
 // TODO replace
-TEST_F(TestFixture, SomeTestName) {}
+TEST_F(TestFixture, SomeTestName7) {}
 
 }  // namespace
diff --git a/test/coverage/utils/base64Test.cpp b/test/coverage/utils/base64Test.cpp
index 34fec4d..e362338 100644
--- a/test/coverage/utils/base64Test.cpp
+++ b/test/coverage/utils/base64Test.cpp
@@ -33,6 +33,6 @@ protected:
 };
 
 // TODO replace
-TEST_F(TestFixture, SomeTestName) {}
+TEST_F(TestFixture, SomeTestName4) {}
 
 }  // namespace
diff --git a/test/extra/NotificationTest.cpp b/test/extra/NotificationTest.cpp
index 54e772d..6f90e21 100644
--- a/test/extra/NotificationTest.cpp
+++ b/test/extra/NotificationTest.cpp
@@ -32,6 +32,6 @@ protected:
 };
 
 // TODO replace
-TEST_F(TestFixture, SomeTestName) {}
+TEST_F(TestFixture, SomeTestName0) {}
 
 }  // namespace
diff --git a/test/extra/PublisherSubscriberTest.cpp b/test/extra/PublisherSubscriberTest.cpp
index 54e772d..916f8d3 100644
--- a/test/extra/PublisherSubscriberTest.cpp
+++ b/test/extra/PublisherSubscriberTest.cpp
@@ -32,6 +32,6 @@ protected:
 };
 
 // TODO replace
-TEST_F(TestFixture, SomeTestName) {}
+TEST_F(TestFixture, SomeTestName1) {}
 
 }  // namespace
diff --git a/test/extra/RpcClientServerTest.cpp b/test/extra/RpcClientServerTest.cpp
index 54e772d..fcdea65 100644
--- a/test/extra/RpcClientServerTest.cpp
+++ b/test/extra/RpcClientServerTest.cpp
@@ -32,6 +32,6 @@ protected:
 };
 
 // TODO replace
-TEST_F(TestFixture, SomeTestName) {}
+TEST_F(TestFixture, SomeTestName2) {}
 
 }  // namespace
