diff --git a/test/coverage/utils/CallbackConnectionTest.cpp b/test/coverage/utils/CallbackConnectionTest.cpp
index a0ae5fb..0b06723 100644
--- a/test/coverage/utils/CallbackConnectionTest.cpp
+++ b/test/coverage/utils/CallbackConnectionTest.cpp
@@ -446,7 +446,7 @@ TEST_F(CallbackTest, ReturnValuesAreMoved) {
 		auto result = callable();
 		EXPECT_TRUE(result);
 		EXPECT_EQ(original_string_location, (*result).data());
-		EXPECT_EQ(expected_capacity, (*result).capacity());
+		EXPECT_LE(expected_capacity, (*result).capacity());
 		// Just to be safe, check our assumptions about copies vs moves. The
 		// a_copy variable should hold a copy of the original string, this time
 		// with a different pointer and capacity.
