diff --git a/tests/CMakeLists.txt b/tests/CMakeLists.txt
index fdce19d..abe8b1d 100644
--- a/tests/CMakeLists.txt
+++ b/tests/CMakeLists.txt
@@ -31,7 +31,7 @@ function(add_test_instance file mode lib option)
 	target_link_libraries(${target} PUBLIC ${lib})
 	set_property(TARGET ${target} PROPERTY LANGUAGE CXX)
 	set_property(TARGET ${target} PROPERTY CXX_STANDARD 17)
-	if (option STREQUAL "Router")
+	if (option STREQUAL "Router" AND NOT CMAKE_CROSSCOMPILING)
 		add_test(NAME "test_${target}" COMMAND bash ${CMAKE_CURRENT_BINARY_DIR}/run_with_router.sh ${target} ${zenoh_router_branch})
 	else()
 		add_test(NAME "test_${target}" COMMAND ${target})
