diff --git a/googletest/cmake/internal_utils.cmake b/googletest/cmake/internal_utils.cmake
index 8c1f9ba9..f70ee36e 100644
--- a/googletest/cmake/internal_utils.cmake
+++ b/googletest/cmake/internal_utils.cmake
@@ -54,12 +54,14 @@ macro(config_compiler_and_linker)
   # Note: pthreads on MinGW is not supported, even if available
   # instead, we use windows threading primitives
   unset(GTEST_HAS_PTHREAD)
-  if (NOT gtest_disable_pthreads AND NOT MINGW)
-    # Defines CMAKE_USE_PTHREADS_INIT and CMAKE_THREAD_LIBS_INIT.
-    set(THREADS_PREFER_PTHREAD_FLAG ON)
-    find_package(Threads)
-    if (CMAKE_USE_PTHREADS_INIT)
-      set(GTEST_HAS_PTHREAD ON)
+  if(!QNX)
+    if (NOT gtest_disable_pthreads AND NOT MINGW)
+      # Defines CMAKE_USE_PTHREADS_INIT and CMAKE_THREAD_LIBS_INIT.
+      set(THREADS_PREFER_PTHREAD_FLAG ON)
+      find_package(Threads)
+      if (CMAKE_USE_PTHREADS_INIT)
+        set(GTEST_HAS_PTHREAD ON)
+      endif()
     endif()
   endif()
 
