From 5a47eb56b99262ed87ca55cbb44ac4e84864500b Mon Sep 17 00:00:00 2001
From: Yongxin Dai <yodai@blackberry.com>
Date: Wed, 12 Feb 2025 15:45:16 -0500
Subject: [PATCH] fix OPENSSL_cpuid_setup for sdp7.1

---
 crypto/cpu_aarch64_qnx.c | 10 ++++++++++
 1 file changed, 10 insertions(+)

diff --git a/crypto/cpu_aarch64_qnx.c b/crypto/cpu_aarch64_qnx.c
index fd3dfda0a..cd5e6b35c 100644
--- a/crypto/cpu_aarch64_qnx.c
+++ b/crypto/cpu_aarch64_qnx.c
@@ -27,21 +27,31 @@ void OPENSSL_cpuid_setup(void) {
 
   OPENSSL_armcap_P |= ARMV7_NEON;
 
+#if defined(AARCH64_CPU_FLAG_AES)
   if (hwcap & AARCH64_CPU_FLAG_AES) {
     OPENSSL_armcap_P |= ARMV8_AES;
   }
+#endif
+#if defined(AARCH64_CPU_FLAG_PMULL)
   if (hwcap & AARCH64_CPU_FLAG_PMULL) {
     OPENSSL_armcap_P |= ARMV8_PMULL;
   }
+#endif
+#if defined(AARCH64_CPU_FLAG_SHA1)
   if (hwcap & AARCH64_CPU_FLAG_SHA1) {
     OPENSSL_armcap_P |= ARMV8_SHA1;
   }
+#endif
+#if defined(AARCH64_CPU_FLAG_SHA256)
   if (hwcap & AARCH64_CPU_FLAG_SHA256) {
     OPENSSL_armcap_P |= ARMV8_SHA256;
   }
+#endif
+#if defined(AARCH64_CPU_FLAG_SHA512)
   if (hwcap & AARCH64_CPU_FLAG_SHA512) {
     OPENSSL_armcap_P |= ARMV8_SHA512;
   }
+#endif
 }
 
 #endif  // OPENSSL_AARCH64 && OPENSSL_FUCHSIA && !OPENSSL_STATIC_ARMCAP
-- 
2.34.1

