From 71b55f847ba44314ead8a1203e065fb592721884 Mon Sep 17 00:00:00 2001
From: Yongxin Dai <yodai@blackberry.com>
Date: Wed, 12 Feb 2025 21:02:51 -0500
Subject: [PATCH] fix template deduction

---
 ssl/tls_record.cc | 7 +++++++
 1 file changed, 7 insertions(+)

diff --git a/ssl/tls_record.cc b/ssl/tls_record.cc
index 685b78ae9..62ab7b8eb 100644
--- a/ssl/tls_record.cc
+++ b/ssl/tls_record.cc
@@ -246,9 +246,16 @@ ssl_open_record_t tls_open_record(SSL *ssl, uint8_t *out_type,
   *out_consumed = in.size() - CBS_len(&cbs);
 
   // In TLS 1.3, during the handshake, skip ChangeCipherSpec records.
+#if defined(__QNX__) && __QNX__ < 800
+  const uint8_t array[] = {SSL3_MT_CCS};
+  if (ssl->s3->version != 0 && ssl_protocol_version(ssl) >= TLS1_3_VERSION &&
+      SSL_in_init(ssl) && type == SSL3_RT_CHANGE_CIPHER_SPEC &&
+      Span<const uint8_t>(body) == Span<const uint8_t>(array)) {
+#else
   if (ssl->s3->version != 0 && ssl_protocol_version(ssl) >= TLS1_3_VERSION &&
       SSL_in_init(ssl) && type == SSL3_RT_CHANGE_CIPHER_SPEC &&
       Span<const uint8_t>(body) == Span<const uint8_t>({SSL3_MT_CCS})) {
+#endif
     ssl->s3->empty_record_count++;
     if (ssl->s3->empty_record_count > kMaxEmptyRecords) {
       OPENSSL_PUT_ERROR(SSL, SSL_R_TOO_MANY_EMPTY_FRAGMENTS);
-- 
2.34.1

