From 77c66dac86bcb0a21a0fe6a4ea2ea576043d6c27 Mon Sep 17 00:00:00 2001
From: Yongxin Dai <yodai@qnx.com>
Date: Thu, 22 May 2025 08:57:19 -0400
Subject: [PATCH] define WEBRTC_POSIX WEBRTC_QNX

---
 BUILD.gn | 7 ++++++-
 1 file changed, 6 insertions(+), 1 deletion(-)

diff --git a/BUILD.gn b/BUILD.gn
index 373e029de9..63d603b35f 100644
--- a/BUILD.gn
+++ b/BUILD.gn
@@ -206,9 +206,14 @@ config("common_inherited_config") {
       target_gen_dir,
     ]
   }
-  if (is_posix || is_fuchsia) {
+  if (is_posix || is_fuchsia || is_qnx) {
     defines += [ "WEBRTC_POSIX" ]
   }
+  if (is_qnx) {
+    defines += [
+      "WEBRTC_QNX",
+    ]
+  }
   if (is_ios) {
     defines += [
       "WEBRTC_MAC",
-- 
2.34.1

