From e4bc42bf9e5aeb276283fa0bee4240ee250a0f80 Mon Sep 17 00:00:00 2001
From: Yongxin Dai <yodai@blackberry.com>
Date: Tue, 18 Feb 2025 14:59:09 -0500
Subject: [PATCH] exclusive fix for SDP7.1: filter out undefined TCP flags

---
 rtc_base/physical_socket_server.cc | 8 ++++++++
 1 file changed, 8 insertions(+)

diff --git a/rtc_base/physical_socket_server.cc b/rtc_base/physical_socket_server.cc
index 042dc15b4a..e16666b9ac 100644
--- a/rtc_base/physical_socket_server.cc
+++ b/rtc_base/physical_socket_server.cc
@@ -756,7 +756,9 @@ int PhysicalSocket::TranslateOption(Option opt, int* slevel, int* sopt) {
         *sopt = IPV6_RECVTCLASS;
       } else {
         *slevel = IPPROTO_IP;
+#if !defined(WEBRTC_QNX) || __QNX__ >= 800
         *sopt = IP_RECVTOS;
+#endif
       }
       break;
 #else
@@ -771,19 +773,25 @@ int PhysicalSocket::TranslateOption(Option opt, int* slevel, int* sopt) {
       break;
     case OPT_TCP_KEEPCNT:
       *slevel = IPPROTO_TCP;
+#if !defined(WEBRTC_QNX) || __QNX__ >= 800
       *sopt = TCP_KEEPCNT;
+#endif
       break;
     case OPT_TCP_KEEPIDLE:
       *slevel = IPPROTO_TCP;
 #if !defined(WEBRTC_MAC)
+#if !defined(WEBRTC_QNX) || __QNX__ >= 800
       *sopt = TCP_KEEPIDLE;
+#endif
 #else
       *sopt = TCP_KEEPALIVE;
 #endif
       break;
     case OPT_TCP_KEEPINTVL:
       *slevel = IPPROTO_TCP;
+#if !defined(WEBRTC_QNX) || __QNX__ >= 800
       *sopt = TCP_KEEPINTVL;
+#endif
       break;
     case OPT_TCP_USER_TIMEOUT:
 #if defined(WEBRTC_LINUX) || defined(WEBRTC_ANDROID)
-- 
2.34.1

