From 2f4b707c44afcdc0311462999fdeb66420abe8c1 Mon Sep 17 00:00:00 2001
From: Yongxin Dai <yodai@qnx.com>
Date: Tue, 18 Mar 2025 08:56:54 -0400
Subject: [PATCH] implement DetectNumberOfCores() for QNX

---
 system_wrappers/source/cpu_info.cc | 4 ++++
 1 file changed, 4 insertions(+)

diff --git a/system_wrappers/source/cpu_info.cc b/system_wrappers/source/cpu_info.cc
index eff720371a..2f4117e4bb 100644
--- a/system_wrappers/source/cpu_info.cc
+++ b/system_wrappers/source/cpu_info.cc
@@ -18,6 +18,8 @@
 #include <sys/sysctl.h>
 #elif defined(WEBRTC_FUCHSIA)
 #include <zircon/syscalls.h>
+#elif defined(WEBRTC_QNX)
+#include <sys/syspage.h>
 #endif
 
 #include "rtc_base/logging.h"
@@ -45,6 +47,8 @@ static int DetectNumberOfCores() {
   }
 #elif defined(WEBRTC_FUCHSIA)
   number_of_cores = zx_system_get_num_cpus();
+#elif defined(WEBRTC_QNX)
+  number_of_cores = static_cast<int>(_syspage_ptr->num_cpu);
 #else
   RTC_LOG(LS_ERROR) << "No function to get number of cores";
   number_of_cores = 1;
-- 
2.34.1

