From f99dbabc5895cca84ce8d971b2c957631271a8bc Mon Sep 17 00:00:00 2001
From: Yongxin Dai <yodai@blackberry.com>
Date: Sat, 15 Apr 2023 10:48:25 -0400
Subject: [PATCH] changes for QNX in p2p/base

---
 p2p/base/turn_server.cc | 9 +++++++++
 1 file changed, 9 insertions(+)

diff --git a/p2p/base/turn_server.cc b/p2p/base/turn_server.cc
index 494939a306..dd935c6803 100644
--- a/p2p/base/turn_server.cc
+++ b/p2p/base/turn_server.cc
@@ -736,8 +736,13 @@ void TurnServerAllocation::HandleChannelBindRequest(const TurnMessage* msg) {
 
   // Add or refresh this channel.
   if (channel1 == channels_.end()) {
+# if defined(WEBRTC_QNX)
+    channel1 = channels_.insert(
+        channels_.end(), {.pending_delete = webrtc::ScopedTaskSafety(), .id = channel_id, .peer = peer_attr->GetAddress()});
+#else
     channel1 = channels_.insert(
         channels_.end(), {.id = channel_id, .peer = peer_attr->GetAddress()});
+#endif
   } else {
     channel1->pending_delete.reset();
   }
@@ -816,7 +821,11 @@ bool TurnServerAllocation::HasPermission(const rtc::IPAddress& addr) {
 void TurnServerAllocation::AddPermission(const rtc::IPAddress& addr) {
   auto perm = FindPermission(addr);
   if (perm == perms_.end()) {
+#if defined(WEBRTC_QNX)
+    perm = perms_.insert(perms_.end(), {.pending_delete = webrtc::ScopedTaskSafety(), .peer = addr});
+#else
     perm = perms_.insert(perms_.end(), {.peer = addr});
+#endif
   } else {
     perm->pending_delete.reset();
   }
-- 
2.34.1

