diff --git a/meson.build b/meson.build
index f453d748c..82d9fcb5d 100644
--- a/meson.build
+++ b/meson.build
@@ -2268,7 +2268,7 @@ libatomic_test_code = '''
   }'''
 atomic_dep = []
 if cc.links(libatomic_test_code, args : '-latomic', name : 'check for -latomic')
-  atomic_dep = cc.find_library('atomic')
+  atomic_dep = cc.find_library('atomic', required: false)
 endif
 
 # First check in libc, fallback to libintl, and as last chance build
