diff --git a/libkineto/src/ThreadUtil.cpp b/libkineto/src/ThreadUtil.cpp
index 3fdc22b..bf0020f 100644
--- a/libkineto/src/ThreadUtil.cpp
+++ b/libkineto/src/ThreadUtil.cpp
@@ -11,7 +11,11 @@
 #ifndef _WIN32
 #include <pthread.h>
 #include <unistd.h>
+#ifdef __QNX__
+#include <process.h>
+#else
 #include <sys/syscall.h>
+#endif
 #include <sys/types.h>
 #else // _WIN32
 #include <locale>
@@ -58,6 +62,8 @@ int32_t systemThreadId() {
     _sysTid = (int32_t)GetCurrentThreadId();
 #elif defined __FreeBSD__
     syscall(SYS_thr_self, &_sysTid);
+#elif defined __QNX__
+    _sysTid = (int32_t)gettid();
 #else
     _sysTid = (int32_t)syscall(SYS_gettid);
 #endif
