From 9ec20402be1ad92ffabf1d7fe413f13583db8330 Mon Sep 17 00:00:00 2001
From: Yongxin Dai <yodai@qnx.com>
Date: Fri, 27 Jun 2025 10:45:07 -0400
Subject: [QNX LOCAL PATCH] Fix __builtin_is_constant_evaluated() for SDP7

---
 core/fxcrt/numerics/safe_conversions_impl.h | 3 +++
 1 file changed, 3 insertions(+)

diff --git a/core/fxcrt/numerics/safe_conversions_impl.h b/core/fxcrt/numerics/safe_conversions_impl.h
index 1b717ca18..b59d0e636 100644
--- a/core/fxcrt/numerics/safe_conversions_impl.h
+++ b/core/fxcrt/numerics/safe_conversions_impl.h
@@ -88,6 +88,9 @@ constexpr typename std::make_unsigned<T>::type SafeUnsignedAbs(T value) {
 
 // TODO(jschuh): Switch to std::is_constant_evaluated() once C++20 is supported.
 // Alternately, the usage could be restructured for "consteval if" in C++23.
+#if defined(__QNX__) && __QNX__ < 800
+extern bool __builtin_is_constant_evaluated (void);
+#endif
 #define IsConstantEvaluated() (__builtin_is_constant_evaluated())
 
 // TODO(jschuh): Debug builds don't reliably propagate constants, so we restrict
-- 
2.34.1

