From d34d20bbee874a79279f01d02ac7c87cff284f95 Mon Sep 17 00:00:00 2001
From: Yongxin Dai <yodai@qnx.com>
Date: Mon, 23 Jun 2025 15:49:33 -0400
Subject: [PATCH] Add '/qnx_build' to .gitignore

---
 .gitignore | 1 +
 1 file changed, 1 insertion(+)

diff --git a/.gitignore b/.gitignore
index 1ffe3ee12..0ab93635f 100644
--- a/.gitignore
+++ b/.gitignore
@@ -3,6 +3,7 @@
 /buildtools
 /compile_commands.json
 /out
+/qnx_build
 /testing/corpus
 /testing/scripts/rust
 /third_party/abseil-cpp
-- 
2.34.1

