From 6d884ffd88a9c687ad9ce98b451f9f5a484457c9 Mon Sep 17 00:00:00 2001
From: Yongxin Dai <yodai@qnx.com>
Date: Sun, 22 Jun 2025 22:49:22 -0400
Subject: [PATCH] Implement PathService::GetExecutableDir for QNX

---
 testing/utils/path_service.cpp | 17 +++++++++++++++++
 1 file changed, 17 insertions(+)

diff --git a/testing/utils/path_service.cpp b/testing/utils/path_service.cpp
index 162687493..abb08019c 100644
--- a/testing/utils/path_service.cpp
+++ b/testing/utils/path_service.cpp
@@ -11,6 +11,12 @@
 #elif defined(__APPLE__)
 #include <mach-o/dyld.h>
 #include <sys/stat.h>
+#include <unistd.h>
+#elif defined(__QNX__)
+#include <fcntl.h>
+#include <limits.h>
+#include <sys/stat.h>
+#include <sys/types.h>
 #else  // Linux
 #include <linux/limits.h>
 #include <sys/stat.h>
@@ -81,6 +87,17 @@ bool PathService::GetExecutableDir(std::string* path) {
   path->resize(path_length - 1);
   if (_NSGetExecutablePath(&((*path)[0]), &path_length))
     return false;
+#elif defined(__QNX__)
+  char raw_buf[PATH_MAX];
+  char buf[PATH_MAX];
+  int fd = open("/proc/self/exefile", O_RDONLY);
+  ssize_t count = read(fd, raw_buf, PATH_MAX);
+  close(fd);
+  if (count > 0) {
+    realpath(raw_buf, buf);
+    count = strlen(buf);
+  }
+  *path = std::string(buf, count);
 #else   // Linux
   static const char kProcSelfExe[] = "/proc/self/exe";
   char buf[PATH_MAX];
-- 
2.34.1

