diff --git a/rosidl_typesupport_fastrtps_c/CMakeLists.txt b/rosidl_typesupport_fastrtps_c/CMakeLists.txt
index 314f933..09c67de 100644
--- a/rosidl_typesupport_fastrtps_c/CMakeLists.txt
+++ b/rosidl_typesupport_fastrtps_c/CMakeLists.txt
@@ -60,6 +60,10 @@ ament_export_targets(${PROJECT_NAME})
 
 ament_index_register_resource("rosidl_typesupport_c")
 
+if(QNX)
+  set(BUILD_TESTING ON)
+endif()
+
 if(BUILD_TESTING)
   find_package(ament_cmake_gtest REQUIRED)
   find_package(ament_lint_auto REQUIRED)
@@ -105,6 +109,10 @@ if(BUILD_TESTING)
 
   find_package(ament_cmake_pytest REQUIRED)
   ament_add_pytest_test(test_cli_extension test/test_cli_extension.py)
+
+  if(QNX)
+    install(DIRECTORY ${CMAKE_BINARY_DIR}/ DESTINATION test/rosidl_typesupport_fastrtps_cpp)
+  endif()
 endif()
 
 ament_package(
diff --git a/rosidl_typesupport_fastrtps_cpp/CMakeLists.txt b/rosidl_typesupport_fastrtps_cpp/CMakeLists.txt
index d42ce68..8de21d9 100644
--- a/rosidl_typesupport_fastrtps_cpp/CMakeLists.txt
+++ b/rosidl_typesupport_fastrtps_cpp/CMakeLists.txt
@@ -64,6 +64,10 @@ ament_export_targets(${PROJECT_NAME})
 
 ament_index_register_resource("rosidl_typesupport_cpp")
 
+if(QNX)
+  set(BUILD_TESTING ON)
+endif()
+
 if(BUILD_TESTING)
   find_package(ament_cmake_gtest REQUIRED)
   find_package(ament_lint_auto REQUIRED)
@@ -111,6 +115,10 @@ if(BUILD_TESTING)
 
   find_package(ament_cmake_pytest REQUIRED)
   ament_add_pytest_test(test_cli_extension test/test_cli_extension.py)
+
+  if(QNX)
+    install(DIRECTORY ${CMAKE_BINARY_DIR}/ DESTINATION test/rosidl_typesupport_fastrtps_cpp)
+  endif()
 endif()
 
 ament_package(
