diff --git a/rosidl_runtime_c/CMakeLists.txt b/rosidl_runtime_c/CMakeLists.txt
index 33baf9b..dc4bc54 100644
--- a/rosidl_runtime_c/CMakeLists.txt
+++ b/rosidl_runtime_c/CMakeLists.txt
@@ -53,6 +53,10 @@ ament_export_targets(${PROJECT_NAME})
 
 ament_index_register_resource("rosidl_runtime_packages")
 
+if(QNX)
+  set(BUILD_TESTING ON)
+endif()
+
 if(BUILD_TESTING)
   find_package(ament_lint_auto REQUIRED)
   find_package(performance_test_fixture REQUIRED)
@@ -108,6 +112,10 @@ if(BUILD_TESTING)
   if(TARGET benchmark_string_conversion)
     target_link_libraries(benchmark_string_conversion ${PROJECT_NAME})
   endif()
+
+  if(QNX)
+    install(DIRECTORY ${CMAKE_BINARY_DIR}/ DESTINATION test/rosidl_runtime_c)
+  endif()
 endif()
 
 install(
