diff --git a/CMakeLists.txt b/CMakeLists.txt
index 7fa4285..aef83f7 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -91,7 +91,11 @@ ELSE()
 ENDIF()
 
 ADD_LIBRARY(pthreadpool_interface INTERFACE)
-TARGET_INCLUDE_DIRECTORIES(pthreadpool_interface INTERFACE include)
+TARGET_INCLUDE_DIRECTORIES(pthreadpool_interface 
+ INTERFACE 
+   "$<BUILD_INTERFACE:${PROJECT_SOURCE_DIR}/include>"
+   "$<INSTALL_INTERFACE:include>"
+)
 IF(NOT PTHREADPOOL_ALLOW_DEPRECATED_API)
   TARGET_COMPILE_DEFINITIONS(pthreadpool_interface INTERFACE PTHREADPOOL_NO_DEPRECATED_API=1)
 ENDIF()
@@ -163,7 +167,8 @@ IF(NOT TARGET fxdiv)
 ENDIF()
 TARGET_LINK_LIBRARIES(pthreadpool PRIVATE fxdiv)
 
-INSTALL(TARGETS pthreadpool
+INSTALL(TARGETS pthreadpool pthreadpool_interface
+  EXPORT tensorflow-liteTargets
   LIBRARY DESTINATION ${CMAKE_INSTALL_LIBDIR}
   ARCHIVE DESTINATION ${CMAKE_INSTALL_LIBDIR})
 
