From 406d8bfffae59e601ec0b6f823aac2b10337b38e Mon Sep 17 00:00:00 2001
From: Yongxin Dai <yodai@qnx.com>
Date: Wed, 9 Apr 2025 12:53:56 -0400
Subject: [PATCH] declare qnx_sdp_root

---
 config/qnx/config.gni | 6 ++++++
 1 file changed, 6 insertions(+)

diff --git a/config/qnx/config.gni b/config/qnx/config.gni
index bbb8df4ba..93af40b18 100644
--- a/config/qnx/config.gni
+++ b/config/qnx/config.gni
@@ -1,6 +1,12 @@
+_qnx_sdp_root = getenv("QNX_TARGET")
 _is_iosnd = exec_script("check_iosnd_header.py", [], "trim string")
 _is_iosock = exec_script("check_iosock_header.py", [], "trim string")
 declare_args() {
+  if (_qnx_sdp_root == "") {
+    error("QNX SDP environment variables not set, please run qnxsdp-env.sh first")
+  } else {
+    qnx_sdp_root = _qnx_sdp_root
+  }
   if (_is_iosnd == "True") {
     rtc_qnx_use_io_snd = true
   } else {
-- 
2.34.1

