From 9ccab5d247d697dd1df98cd6ca930b0f4a0b4e70 Mon Sep 17 00:00:00 2001
From: Yongxin Dai <yodai@qnx.com>
Date: Fri, 20 Jun 2025 11:23:08 -0400
Subject: [PATCH] define ARMV8_OS_QNX

---
 BUILD.gn | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/BUILD.gn b/BUILD.gn
index 378bd0d..419b9ff 100644
--- a/BUILD.gn
+++ b/BUILD.gn
@@ -138,6 +138,8 @@ if (use_arm_neon_optimizations) {
       defines += [ "ARMV8_OS_FUCHSIA" ]
     } else if (is_win) {
       defines += [ "ARMV8_OS_WINDOWS" ]
+    } else if (is_qnx) {
+      defines += [ "ARMV8_OS_QNX" ]
     } else {
       assert(false, "Unsupported ARM OS")
     }
-- 
2.34.1

