From 70e8fbdb36b19f1099cfb03422e19caf6abac468 Mon Sep 17 00:00:00 2001
From: Yongxin Dai <yodai@qnx.com>
Date: Sun, 22 Jun 2025 22:51:03 -0400
Subject: [PATCH] Changes for QNX

---
 cpu_features.c | 13 +++++++++++--
 1 file changed, 11 insertions(+), 2 deletions(-)

diff --git a/cpu_features.c b/cpu_features.c
index 34ae7b9..d266ace 100644
--- a/cpu_features.c
+++ b/cpu_features.c
@@ -39,7 +39,8 @@ int ZLIB_INTERNAL riscv_cpu_enable_vclmul = 0;
 #ifndef CPU_NO_SIMD
 
 #if defined(ARMV8_OS_ANDROID) || defined(ARMV8_OS_LINUX) || \
-    defined(ARMV8_OS_FUCHSIA) || defined(ARMV8_OS_IOS)
+    defined(ARMV8_OS_FUCHSIA) || defined(ARMV8_OS_IOS) || \
+    defined(ARMV8_OS_QNX)
 #include <pthread.h>
 #endif
 
@@ -56,6 +57,8 @@ int ZLIB_INTERNAL riscv_cpu_enable_vclmul = 0;
 #include <windows.h>
 #elif defined(ARMV8_OS_IOS)
 #include <sys/sysctl.h>
+#elif defined(ARMV8_OS_QNX)
+#include <sys/syspage.h>
 #elif !defined(_MSC_VER)
 #include <pthread.h>
 #else
@@ -69,7 +72,7 @@ static void _cpu_check_features(void);
 #if defined(ARMV8_OS_ANDROID) || defined(ARMV8_OS_LINUX) || \
     defined(ARMV8_OS_MACOS) || defined(ARMV8_OS_FUCHSIA) || \
     defined(X86_NOT_WINDOWS) || defined(ARMV8_OS_IOS) || \
-    defined(RISCV_RVV)
+    defined(RISCV_RVV) || defined(ARMV8_OS_QNX)
 #if !defined(ARMV8_OS_MACOS)
 // _cpu_check_features() doesn't need to do anything on mac/arm since all
 // features are known at build time, so don't call it.
@@ -144,6 +147,12 @@ static void _cpu_check_features(void)
     len = sizeof(val);
     arm_cpu_enable_pmull = sysctlbyname("hw.optional.arm.FEAT_PMULL", &val, &len, 0, 0) == 0
                && val != 0;
+#elif defined(ARMV8_OS_QNX)
+    // TODO: yodai HWCAP_CRC32 has already defined in SDP devs's machine/elf.h,
+    // probably will be supported soon.
+    arm_cpu_enable_crc32 = 0;
+    uint32_t hwcap = SYSPAGE_ENTRY(cpuinfo)->flags;
+    arm_cpu_enable_pmull = hwcap & AARCH64_CPU_FLAG_PMULL;
 #endif
 }
 #endif
-- 
2.34.1

